File: game/projectilewars/scripts/vscripts/modifier_silence.lua
require("typescript_lualib")
modifier_silence_basic = {}
modifier_silence_basic.__index = modifier_silence_basic
function modifier_silence_basic.new(construct, ...)
    local instance = setmetatable({}, modifier_silence_basic)
    if construct and modifier_silence_basic.constructor then modifier_silence_basic.constructor(instance, ...) end
    return instance
end
function modifier_silence_basic.IsHidden(self)
    return false
end
function modifier_silence_basic.IsPurgable(self)
    return true
end
function modifier_silence_basic.GetAttributes(self)
    return MODIFIER_ATTRIBUTE_MULTIPLE
end
function modifier_silence_basic.OnCreated(self,kv)
    if IsServer() then
        local parent = CDOTA_Buff.GetParent(self)

        local silenceIndices = TS_ITE(kv.silenceIndices and (#kv.silenceIndices<1),function() return kv.silenceIndices end,function() return {1,2,3,4,5,6} end)

        local abil = CDOTA_BaseNPC.GetCurrentActiveAbility(CDOTA_Buff.GetCaster(self))

        if base_ability.CanBeSilenced(abil) and (TS_indexOf(silenceIndices, CDOTABaseAbility.GetAbilityIndex(abil))~=-1) then
            CDOTA_BaseNPC.Interrupt(CDOTA_Buff.GetParent(self))
        end
        local silencableAbilities = {}

        for i=0,8-1,1 do
            local ability = CDOTA_BaseNPC.GetAbilityByIndex(CDOTA_Buff.GetParent(self),i)

            if (ability and base_ability.CanBeSilenced(ability)) and (TS_indexOf(silenceIndices, i)~=-1) then
                base_ability.SetSilenceEndTime(ability,CDOTA_Buff.GetRemainingTime(self))
                table.insert(silencableAbilities, i)
            end
        end
        CCustomGameEventManager.Send_ServerToPlayer(CustomGameEventManager,CDOTA_BaseNPC.GetPlayerOwner(CDOTA_Buff.GetParent(self)),"hero_silence_created",{silenceIndex=silencableAbilities})
    end
end
function modifier_silence_basic.OnDestroy(self)
    if IsServer() then
        local parent = CDOTA_Buff.GetParent(self)

        local silenceIndices = {}

        for i=0,8-1,1 do
            local ability = CDOTA_BaseNPC.GetAbilityByIndex(parent,i)

            if ability and not base_ability.IsSilenced(ability) then
                table.insert(silenceIndices, i)
            end
        end
        CCustomGameEventManager.Send_ServerToPlayer(CustomGameEventManager,CDOTA_BaseNPC.GetPlayerOwner(CDOTA_Buff.GetCaster(self)),"hero_silence_removed",{unsilenceIndex=silenceIndices})
    end
end
function modifier_silence_basic.GetEffectName(self)
    return "particles/generic_gameplay/generic_silenced.vpcf"
end
function modifier_silence_basic.GetEffectAttachType(self)
    return PATTACH_OVERHEAD_FOLLOW
end

File: game/projectilewars/scripts/vscripts/modifier_silence.ts
class modifier_silence_basic extends CDOTA_Modifier_Lua {

  IsHidden() {return false}
  IsPurgable() {return true}
  GetAttributes() { return DOTAModifierAttribute_t.MODIFIER_ATTRIBUTE_MULTIPLE}
  OnCreated(kv:{silenceIndices:number[]}) {
    if (IsServer()) {
      let parent = this.GetParent()
      let silenceIndices = kv.silenceIndices && kv.silenceIndices.length < 1 ? kv.silenceIndices : [1,2,3,4,5,6]
      let abil = this.GetCaster().GetCurrentActiveAbility() as base_ability
      // Interrupt
      if (abil.CanBeSilenced() && silenceIndices.indexOf(abil.GetAbilityIndex()) != -1) {
        this.GetParent().Interrupt()
      }
      //Activate overlay
      let silencableAbilities = []
      for (let i = 0;i <8;i++) {
        let ability = this.GetParent().GetAbilityByIndex(i) as base_ability
        if (ability && ability.CanBeSilenced() && silenceIndices.indexOf(i) != -1) {
          ability.SetSilenceEndTime(this.GetRemainingTime())
          silencableAbilities.push(i)
        }
      }
      CustomGameEventManager.Send_ServerToPlayer(this.GetParent().GetPlayerOwner(),"hero_silence_created",{silenceIndex:silencableAbilities})
    } 
  }

  OnDestroy() {
    if (IsServer()) {
      let parent = this.GetParent()
      let silenceIndices = []
      for (let i=0;i< 8;i++) {
        let ability = parent.GetAbilityByIndex(i) as base_ability
        if (ability && !ability.IsSilenced()) {
          silenceIndices.push(i)
        }
      }
      //Remove overlay
      CustomGameEventManager.Send_ServerToPlayer(this.GetCaster().GetPlayerOwner(),"hero_silence_removed",{unsilenceIndex:silenceIndices})
    }
  }

  //IsSilenceDebuff:true

  GetEffectName() {
    return "particles/generic_gameplay/generic_silenced.vpcf"
  }
  GetEffectAttachType() {
    return ParticleAttachment_t.PATTACH_OVERHEAD_FOLLOW
  }

}
